sstables/sstable_set.cc

```diff
@@ -265,7 +265,12 @@ partitioned_sstable_set::query(const dht::partition_range& range) const {
 }
 
 bool partitioned_sstable_set::store_as_unleveled(const shared_sstable& sst) const {
-    return _use_level_metadata && sst->get_sstable_level() == 0;
+    // When a sstable spans most of the entire token range, we'll store it in a
+    // vector, to avoid triggering quadratic space complexity in the interval map,
+    // since many of such sstables would have presence on almost all intervals.
+    static constexpr float unleveled_threshold = 0.85f;
+    auto sst_tr = dht::token_range(sst->get_first_decorated_key().token(), sst->get_last_decorated_key().token());
+    return dht::overlap_ratio(_token_range, sst_tr) >= unleveled_threshold;
```

Comment:
Can we do something simpler instead and rely on sstable origin?
All sstables flushed from memtable would be considered unleveled and not inserted into the sstable set.
SStables flushed from memtables usually span the whole ring/tablet. After compaction, they either have constrained range (ICS) or are larger and we there are fewer of them so we can add them to the set anyway.

Reply:
We thought about that. But during a flush storm, compaction output is still a small sized sstable spanning almost all the range. Using origin would miss this.